sdks/python/apache_beam/runners/common.py

```diff
@@ -735,7 +733,44 @@ def invoke_process(self,
     else:
       self._invoke_process_per_window(
           windowed_value, additional_args, additional_kwargs)
-    return None
+    return residuals
+
+  def _should_process_window_for_sdf(
+      self,
+      windowed_value, # type: WindowedValue
+      additional_kwargs,
+      window_index=None, # type: Optional[int]
+  ):
+    restriction_tracker = self.invoke_create_tracker(self.restriction)
```

Comment:
Would it be better to check whether `False` should be returned first then create `restriction_tracker` and `watermark_estimator`?

And the function itself looks kind of ambiguous to me. The function is used no matter whether the window is observed to prepare the params for processing sdf. Is it possible to separate the check and creating params part?

Reply:
I don't want to hold the splitting lock while the user methods are being invoked.